### PR TITLE
Intro JSONStreamReader and JSONify2

### PR DESCRIPTION
**Pull request by jf-tech**

### CODE REVIEW

File: idr/.snapshots/TestJSONify2XML [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (wangjia007bond) [inline idr/.snapshots/TestJSONify2XML:1]: is this a special json file, why the key are like this "xmlns:a", "#attributes"

File: idr/marshal2.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+		return b
+	case IsJSONValueNull(n):
+		return nil
+	default:
+		return n.Data
+	}
+}
+
+// J2NodeToInterface converts *Node into an interface{} that, once marshaled, is JSON friendly.
+func J2NodeToInterface(n *Node) interface{} {
```
Comment (wangjia007bond) [inline idr/marshal2.go:134]: looks like we want to convert xml to json? why we want do this?

File: idr/.snapshots/TestJSONify2XML (hunk 1 above)
Comment (jf-tech) [reply idr/.snapshots/TestJSONify2XML:1]: All files in `.snapshots` folders are, well, snapshot files from tests.

JSON does allow these chars in property names. And you can access them in JavaScript like this:

`var n = JSON.parse(...); console.log(n["#attributes]);`

File: idr/marshal2.go (hunk 2 above)
Comment (jf-tech) [reply idr/marshal2.go:134]: There was demand that people wanted to simple copy a section of json input to output, so we introduced this JSONify2 (under a different name) . However the code was written such that it works for all Node trees from different input formats.

Not only that, now we have `javascript` custom function, which we pass in a special argument `_node` that is JSON serialized Node of the current context, so schema writer can do whatever custom processing they see to fit. In this case it's even more important to pass in a json doc translated from all supported input formats (not just JSON/XML, but all input formats)

File: idr/jsonnode.go [hunk 3]
```diff
@@ -1,33 +1,73 @@
 package idr
 
+// JSONType is the type of a JSON-specific Node.
+// Note multiple JSONType can be bit-wise OR'ed together.
 type JSONType uint
 
 const (
+	// JSONRoot is the type for the root Node in a JSON IDR tree.
 	JSONRoot JSONType = 1 << iota
+	// JSONRoot is the type for a Node in a JSON IDR tree whose value is an object.
```
Comment (wangjia007bond) [inline idr/jsonnode.go:10]: here the comment seems should be `JSONObj`

File: idr/marshal2.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		for c := n.FirstChild; c != nil; c = c.NextSibling {
+			name := j2NodeName(c)
+			value := J2NodeToInterface(c)
+			if c.Type == ElementNode {
+				// Both in XML and EDI, we can potentially have something like the following:
+				//   <abc>
+				//     <efg>1</abc>
+				//     <efg>2</abc>
+				//     <xyz>3</efg>
+				//   </abc>
```
Comment (wangjia007bond) [inline idr/marshal2.go:170]: is this is a correct xml, should we need to deal with this?

File: idr/jsonnode.go (hunk 3 above)
Comment (jf-tech) [reply idr/jsonnode.go]: fixed

File: idr/marshal2.go (hunk 4 above)
Comment (jf-tech) [reply idr/marshal2.go]: fixed

**Review by wangjia007bond — APPROVED** [review]

**Pull request merged**